### ISSUE TITLE
Support container_name in docker-compose file

### ISSUE THREAD

**Issue by krichter722** [testcontainers/testcontainers-java#2472]
`docker-compose` files which contain `container_name` are not supported and the creation of `DockerComposeContainer` fails due to `IllegalStateException`. This has been introduced in #1151 as a quick fix for unintuitive feedback. If I understand correctly, the option of an implementation was however not discarded, but postponed. This issue is now about the implementation.

**Comment by bsideup** [comment]
Hi @krichter722,

Would you like to work on implementing it? Also, do you see it as a blocker for you, or just a nice to have feature?

**Comment by isank** [comment]
Hi @bsideup,

I would like to work on this.

**Comment by reggiemcdonald** [comment]
Has this been finished? I would be interested in working on this if it isn't.

**Comment by bsideup** [comment]
@reggiemcdonald it is still up for grabs 👍

**Comment by swanysimon** [comment]
@kiview pulling the conversation back to the original issue

https://github.com/testcontainers/testcontainers-java/pull/2741#issuecomment-1225513472

I don't really disagree with your comment on the PR, and it lines up with the general messaging from this project, but that's not necessarily why people are using this library. For a lot of us, we're transitioning manual workflows into automated tests. We just want to be able to use the same file we run `docker-compose up` against in automated tests. For the manual workflows, it's a huge PITA to have to go look up the generated name for each of your containers, so we attach container names to them.

It might be worth closing this issue with the same (or similar) comment you put on the PR if you believe names _shouldn't_ be used in compose files, but otherwise I'm happy to talk through my options here.

**Comment by AlHoz** [comment]
I completely agree with @swanysimon. I need this feature only for re-using the compose file used in prod. Without it, I have to maintain two files and keep them synced. How about the middle ground: testcontainers can parse a compose file with container names, but ignore the configuration option when run docker (names are not assigned to containers)?

**Comment by Kamveno** [comment]
I would like to contribute, can we merge?

**Comment by jlengrand** [comment]
There seems to be some pushback as to supporting the `container_name` feature. At the same time, I can imagine use cases where I'm not in total control of the Docker Compose file that is provided to me, say if I want to run a neighbouring microservice. Would it not be possible to at least throw a warning and ignore the parameter (at the risk of the tests failing for the users) instead throwing an exception?

I am asking because the only other option then is to either duplicate the file, or edit it temporarily. I'd even be happy if the feature was behind a flag and had to be specifically activated.

**Comment by commonquail** [comment]
Testcontainer's hard failure in preference to simply ignoring the value is difficult to justify but it is worth pointing out that in the four years that have passed since this issue was created, and five since #1151, Docker Compose has learned to generate [`COMPOSE_PROJECT_NAME`](https://docs.docker.com/compose/environment-variables/envvars/#compose_project_name) from the containing directory's name. Therefore, in the absence of both a `container_name` setting and an explicitly set project name, Docker Compose will produce a  container name of  `<containing directory>-<service name>-<n>` instead of merely `<service name>-<n>`. Consequently, this issue has been effectively ignored into irrelevance.

**Comment by fonfalleh** [comment]
Regardless of if container_name should be supported, the documentation about using compose could mention that they are not supported to save some time and frustration.
https://java.testcontainers.org/modules/docker_compose/

**Comment by prayas075** [comment]
Hi, I would like to work on this issue and implement support for container_name in DockerComposeContainer. 
I’ll fork the repo, make the changes, and submit a PR.

**Comment by ST-DDT** [comment]
I ended up up copying the file before using it:

````java
List<String> lines = Files.readAllLines(new File("docker-compose.yml").toPath());
lines.removeIf(line -> line.includes("container_name:"));
Files.write(new File("docker-compose-temp.yml").toPath(), lines);
````

I ran into more problems later on because of the docker-compose-in-docker, in the end I abandonned this approach.

**Comment by exijn** [comment]
Hi, I’m interested in working on this issue.

I regularly use Docker Compose with Spring Boot, MariaDB, and Redis services, so the need to reuse an existing Compose file containing `container_name` is familiar to me.

I see that the thread has discussed both supporting the field and ignoring it with a warning. Before preparing an implementation, could a maintainer confirm which behavior would currently be acceptable?

Once the expected behavior is confirmed, I can add focused coverage for Compose parsing and service discovery and submit a small PR.

**Comment by ST-DDT** [comment]
Not a maintainer, but I tried ignoring it. 
That didnt work because the files are used as is.
The validation is just added on top and doesnt really impact the final result other than failing on unsupported properties.
(E.g. you would have to create temp files without the property)

So the options are:

- support the properties
- Create  temp files without the property

Both have pros and cons:

- Supporting the properties will cause conflicts with containers created directly via the compose file, unless you also use file level disambiguators for the whole container set.
- Using temp files causes issues with dangling files (post run cleanup) and referenced files that need to be included. E.g. if a compose file links to a docker file.
